### ISSUE TITLE
stream through tar w/o compression?

### ISSUE THREAD

**Issue by jamestwebber** [chanzuckerberg/s3mi#5]
I made a giant tar file without compressing it, just to make S3 transfer faster. `s3cp.py` is great for downloading it, but it'd be even greater if I could simultaneously untar the result. Unfortunately I got an error trying to stream through `s3cp.py path_to_file - | tar xf -`: 

```
File size is 107.3 GB (115159378944 bytes).
Fetching 859 segments.

S3MI tools s3cp

Usage similar to "aws s3 cp", but 10x faster for large files (10+ GB)
on machines that have 10+ Gbps connection to S3.

Examples:

    # Copy from S3 to local file
    s3cp.py s3://my_bucket/my_big_file /mnt/instance_local_path/my_big_file

    # Stream from S3 through tar
    s3cp.py s3://my_bucket/my_big_file.tar.bz2 - | lbzip2 -dc | tar xvf -

License:
    https://github.com/chanzuckerberg/s3mi/blob/master/LICENSE

Traceback (most recent call last):
  File "/home/james/anaconda3/envs/albacore/bin/s3cp.py", line 4, in <module>
    __import__('pkg_resources').run_script('s3mi==0.0.5', 's3cp.py')
  File "/home/james/anaconda3/envs/albacore/lib/python3.6/site-packages/pkg_resources/__init__.py", line 748, in run_script
    self.require(requires)[0].run_script(script_name, ns)
  File "/home/james/anaconda3/envs/albacore/lib/python3.6/site-packages/pkg_resources/__init__.py", line 1517, in run_script
    exec(code, namespace, namespace)
  File "/home/james/anaconda3/envs/albacore/lib/python3.6/site-packages/s3mi-0.0.5-py3.6.egg/EGG-INFO/scripts/s3cp.py", line 209, in <module>
    destination=(len(sys.argv) > 2 and sys.argv[2]) or "-")
  File "/home/james/anaconda3/envs/albacore/lib/python3.6/site-packages/s3mi-0.0.5-py3.6.egg/EGG-INFO/scripts/s3cp.py", line 161, in main
    sys.stdout = os.fdopen(sys.stdout.fileno(), 'ab', FILE_BUFFER_SIZE)
  File "/home/james/anaconda3/envs/albacore/lib/python3.6/os.py", line 1017, in fdopen
    return io.open(fd, *args, **kwargs)
OSError: [Errno 29] Illegal seek
```

Any thoughts? This is with Python 3.6 on Ubuntu 16.

**Comment by boris-dimitrov** [comment]
Thanks for the bug report.   Streaming uncompressed tar is actually a great use case.   I am in the process of rewriting this to make it work much better.   Here is the draft documentation for the new version.

```
s3mi cp s3://huge_file destination

    -- fast 2GB/sec download from S3

    -- may be constrained by destination write bandwidth

        * when the destination is an EBS gp2 volume,
          write bandwidth is only 160 MB/sec

        * RAID can increase that to 1.75GB/sec
          on select instance types [1]


s3mi cat s3://huge_file | some_command

   -- use cases

          * expand uncompressed archives

              s3cat s3://gigabytes.tar | tar xf -

          * stream through a really fast computation

              s3cat s3://gigabytes_of_text | wc -l

   -- do not use for expanding compressed archives

          * typically gated by decompression or
            by destination, not by download

   -- use only when piping through another command

          * s3mi cp is better than s3mi cat
            for direct downloads


s3mi raid volume-name [N] [volume-size]

      Use RAID to overcome destination bandwidth limits.

      Example:

          s3mi raid my_fast_raid 7 214GB

      Creates 7 x 214GB EBS gp2 volumes, RAID0s those together,
      and mounts the set on /mnt/my_fast_raid.

      After the instance is restarted or terminated, the RAID array
      will persist, but will not be mounted.  To remount on either the
      original instance, or on another instance after the original
      instance has been terminated, just rerun the same command

          s3mi raid my_fast_raid 7 214GB

      Optimal configuration:

      The ideal N is the per-instance EBS bandwidth limit [1]
      divided by the per-volume EBS bandwidth limit [2].

      The ideal volume-size is chosen to ensure the per-volume
      bandwidth limit can be met in steady state [2].

      In Dec 2017, the ideal settings are as follows.

          c5.18xlarge with gp2 EBS

              N >= 7
              volume-size >= 214 GB

          i3.16xlarge with gp2 EBS

              N >= 11
              volume-size >= 214 GB
```

**Comment by boris-dimitrov** [comment]
I think in this case for some reason stdout cannot be reopened with a different buffer size;   I will just get rid of that optimization and it should stop it from breaking.

**Comment by boris-dimitrov** [comment]
I can't seem to repro this.   I wonder if it just doesn't happen in the latest version.  Can you try again after `pip install --upgrade` installs version 0.3.0?   If it still happens, I'd like to poke at it on your machine.

**Comment by jamestwebber** [comment]
Seems to work now! Not sure what the issue was before but a lot of things changed.

One note (which is really a separate issue): it would be nice if `s3mi help` or `s3mi --help` worked...

**Comment by boris-dimitrov** [comment]
Added --help support and a couple more niceties in version 0.4.1.
